### PR TITLE
cache, resmgr: expose key evaluation, implement key substitution.

### PR DESCRIPTION
**Pull request by klihub**
Expose evaluating pod and container key references using the newly added `EvalRef()` Evaluable interface. This lets one evaluate values of key references, something which previously was only done as part of evaluating expressions. Also, introduce `Substitute()` which takes a source string and performs key substitution on it wrt. a subject, using the new `EvalRef()`, to resolve key references in plain ('$key') or well-formed ('${key}') notation.

### CODE REVIEW

File: pkg/apis/resmgr/v1alpha1/expression.go [hunk 1]
```diff
@@ -296,9 +301,97 @@ func ResolveRef(subject Evaluable, spec string) (string, bool, error) {
...
+
+			if err := substKeyValue(); err != nil {
+				return "", err
+			}
+
+			buf.Reset()
+
+		case '{':
+			if curly || buf.Len() != 0 {
+				return "", fmt.Errorf("invalid substitution %q, unexpecetd '{' (at %q)",
```
Comment (askervin) [inline pkg/apis/resmgr/v1alpha1/expression.go:360]: s/unexpecetd/unexpected

**Review by askervin — COMMENTED** [review]
This looks great. Just a typo-nit.

**Comment by askervin** [comment]
When implementing the groupBy option (PR #278) on top of this, I wondered if we should scope the `Substitute()` function so that it would not be directly in `resmgr/v1alpha1` scope, but under `resmgr/expression` or something similar.  The reason for considering this is that seeing `resmgr.Substitute()` in the code might not associate with expression handling.

But I'm definitely not requesting this change, only asking about your feelings on the location of the expression part of the API. On the other hand, I would not like to complicate the API with unnecessary sub-namespaces/scopes, unless there is really need for it.

**Comment by klihub** [comment]
> When implementing the groupBy option (PR #278) on top of this, I wondered if we should scope the `Substitute()` function so that it would not be directly in `resmgr/v1alpha1` scope, but under `resmgr/expression` or something similar. The reason for considering this is that seeing `resmgr.Substitute()` in the code might not associate with expression handling.

Hmm, would adding a `Substitute(src string, mustResolve bool) (string, error)` convenience function to containers be of any help in this regard ? It would not add the mental association to expressions, but it would be less intrusive (would not require reshuffling the package tree right now), and it would leave resmgrapi out of the function signature/call altogether, maybe making it more intuitive what the function does just by looking at its signature.

> 
> But I'm definitely not requesting this change, only asking about your feelings on the location of the expression part of the API. On the other hand, I would not like to complicate the API with unnecessary sub-namespaces/scopes, unless there is really need for it.

**Comment by askervin** [comment]
> Hmm, would adding a `Substitute(src string, mustResolve bool) (string, error)` convenience function to containers be of any help in this regard ?

Good idea, I think it would. Then it would be next to `ctr.EvalRef()` for evaluating single references, so it would nicely fit there in my mind at least. And as it takes single evaluable as a parameter, it fits fine there.

> maybe making it more intuitive what the function does just by looking at its signature.

Would `Expand(string)` sound descriptive enough?

**Comment by klihub** [comment]
> > Hmm, would adding a `Substitute(src string, mustResolve bool) (string, error)` convenience function to containers be of any help in this regard ?
> 
> Good idea, I think it would. Then it would be next to `ctr.EvalRef()` for evaluating single references, so it would nicely fit there in my mind at least. And as it takes single evaluable as a parameter, it fits fine there.
> 
> > maybe making it more intuitive what the function does just by looking at its signature.
> 
> Would `Expand(string)` sound descriptive enough?

Yes, probably `Expand` is better than `Substitute`. How about the mustResolve argument ? Should we omit it from the container-level convenience wrapper and let unresolved keys always evaluate to `""` ?

**Comment by klihub** [comment]
@askervin Updated, renaming `Substitute()` to `Expand()` and adding convenience wrappers to pods and containers. The wrappers also require a `mustResolve bool` since I recall you used `Substitute()` with `mustResolve = true` in the initial placement and with `mustResolve = false` in later updates.

**Review by askervin — APPROVED** [review]
LGTM. This is very nice!

**Pull request merged**